zeus/monitor.py

```diff
@@ -0,0 +1,230 @@
+        self.device_id = device_id
+
+        # Spawn the Zeus monitor.
+        if zeus_monitor_log_dir:
+            self._power_csv = f"{zeus_monitor_log_dir}/gpu{self.device_id}.power.csv"
+        else:
+            self._power_csv = tempfile.mkstemp(
+                suffix=f"+gpu{self.device_id}.power.csv"
+            )[1]
+        self._monitor = subprocess.Popen(
+            args=[
+                zeus_monitor_path,
+                self._power_csv,
+                "0",
+                str(zeus_monitor_sleep_ms),
+                str(self.device_id),
+            ],
+            stdin=subprocess.DEVNULL,
+        )
+        self._time_origin = time.monotonic()
+
+        # Make sure the monitor is eventually stopped.
+        def exit_hook():
+            self._monitor.send_signal(signal.SIGINT)
+            time.sleep(2.0)
+            self._monitor.kill()
+
+        atexit.register(exit_hook)
+
+        # Set internal profiling states.
+        self._started_steps = 0
+        self._finished_steps = 0
+        self._profile_start_times: list[float] = []
+        self._profile_end_times = []
+        self._metric_cache = {}
+
+    def start_step(self) -> None:
+        """Mark the beginning of one step."""
+        torch.cuda.synchronize()
+        current_time = time.monotonic()
+        self._started_steps += 1
+        if (
+            self.skip_steps
+            < self._started_steps
+            <= self.skip_steps + self.profile_steps
+        ):
+            self._profile_start_times.append(current_time - self._time_origin)
+
+    def finish_step(self) -> None:
+        """Mark the end of one step."""
+        torch.cuda.synchronize()
+        current_time = time.monotonic()
+        self._finished_steps += 1
+        if (
+            self.skip_steps
+            < self._finished_steps
+            <= self.skip_steps + self.profile_steps
+        ):
+            self._profile_end_times.append(current_time - self._time_origin)
+
+    @contextmanager
+    def step(self) -> Generator[None, None, None]:
+        """Wrap one training step to mark start and finish times."""
+        try:
+            self.start_step()
+            yield
+        finally:
+            self.finish_step()
+
+    @property
+    def is_done(self) -> bool:
+        """Return whether the specified profiling steps are done."""
+        if self._started_steps != self._finished_steps:
+            raise RuntimeError(
+                "`is_done` should be called outside of the code range marked as a 'step'."
+            )
+        return self._finished_steps >= self.skip_steps + self.profile_steps
+
+    @property
+    def total_energy(self) -> float:
+        """Return the total energy consumption of `profile_steps` steps."""
+        try:
+            return self._metric_cache["total_energy"]
+        except KeyError:
+            metric = sum(
+                compute_energy(self._power_csv, start=start, end=end)
```

Comment:
Just my question about the `analyze.energy` function here: we use the relative time for `start` and `end` instead of absolute time. Is it because we try to avoid the inconsistency with monitor's time logging (in C++)?

Reply:
Yes, and also since it's more convenient because negative start/end times can express "wrap arounds".